algebra-core/src/curves/batch_verify.rs

```diff
@@ -0,0 +1,224 @@
+use crate::fields::FpParameters;
+use crate::{
+    cfg_chunks_mut,
+    curves::{batch_bucketed_add_split, BatchGroupArithmeticSlice},
+    AffineCurve, PrimeField, ProjectiveCurve, Vec,
+};
+use num_traits::{identities::Zero, Pow};
+
+use core::fmt;
+#[cfg(feature = "parallel")]
+use rand::thread_rng;
+use rand::Rng;
+
+const MAX_BUCKETS_FOR_FULL_CHECK: usize = 2;
+
+#[cfg(feature = "parallel")]
+use rayon::prelude::*;
+
+#[derive(Debug, Clone)]
+pub struct VerificationError;
+
+impl fmt::Display for VerificationError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "Verification Error. Not in subgroup")
+    }
+}
+
+// Only pass new_security_param if possibly recursing
+fn verify_points<C: AffineCurve, R: Rng>(
+    points: &[C],
+    num_buckets: usize,
+    new_security_param: Option<usize>,
+    rng: &mut R,
+) -> Result<(), VerificationError> {
+    let mut bucket_assign = Vec::with_capacity(points.len());
+    for _ in 0..points.len() {
+        bucket_assign.push(rng.gen_range(0, num_buckets));
+    }
+    let mut buckets = batch_bucketed_add_split(num_buckets, points, &bucket_assign[..], 12);
+
+    // Check that all the buckets belong to the subgroup, either by calling
+    // the batch verify recusively, or by directly checking by multiplying by group order
+    // when the number of buckets is small enough
+    if num_buckets <= MAX_BUCKETS_FOR_FULL_CHECK || new_security_param == None {
+        // We use the batch scalar mul to check the subgroup condition if
+        // there are sufficient number of buckets
+        let verification_failure = if num_buckets >= 4096 {
+            cfg_chunks_mut!(buckets, 4096).for_each(|e| {
+                let length = e.len();
+                e[..].batch_scalar_mul_in_place::<<C::ScalarField as PrimeField>::BigInt>(
+                    &mut vec![<C::ScalarField as PrimeField>::Params::MODULUS.into(); length][..],
+                    1,
+                );
+            });
+            !buckets.iter().all(|&p| p == C::zero())
+        } else {
+            !buckets.iter().all(|&b| {
+                b.mul(<C::ScalarField as PrimeField>::Params::MODULUS) == C::Projective::zero()
+            })
+        };
+        if verification_failure {
+            return Err(VerificationError);
+        }
+    } else {
+        // Since !new_security_param.is_none():
+        let new_security_param = new_security_param.unwrap();
```

Comment:
Seems like this whole branch of code is only for recursion. Let's just remove it and create an issue.